### PR TITLE
PMM-6899 Cluster external connections

### PR DESCRIPTION
**Pull request by percona-csalguero**

### CODE REVIEW

File: deploy/install_eks_operators.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+TOP_DIR=$(git rev-parse --show-toplevel)
+PMM_USER="$(echo -n 'admin' | base64)";
+PMM_PASS="$(echo -n 'admin_password' | base64)";
+KUBECTL_CMD="kubectl --kubeconfig ${HOME}/.kube/config_eks"
+
+# Install the PXC operator
+cat ${TOP_DIR}/deploy/pxc_operator.yaml | ${KUBECTL_CMD} apply -f -
+cat ${TOP_DIR}/deploy/secrets.yaml | sed "s/pmmserver:.*=/pmmserver: ${PMM_PASS}/g" | ${KUBECTL_CMD} apply -f -
```
Comment (BupycHuk) [inline deploy/install_eks_operators.sh:10]: I'm pretty sure that it should be two different secrets for pxc and psmdb.

File: tests/psmdbcluster/psmdb_cluster_test.go [hunk 2]
```diff
@@ -80,31 +85,44 @@ func TestPSMDBClusterAPI(t *testing.T) {
 
 	for _, cluster := range clusters.Clusters {
 		if cluster.Name == name {
-			assert.Equal(t, int32(3), cluster.Params.ClusterSize)
-			assert.Equal(t, int64(1024*1024*1024), cluster.Params.Replicaset.ComputeResources.MemoryBytes)
-			assert.Equal(t, int32(1000), cluster.Params.Replicaset.ComputeResources.CpuM)
+			assert.Equal(t, clusterSize, cluster.Params.ClusterSize)
+			assert.Equal(t, memory, cluster.Params.Replicaset.ComputeResources.MemoryBytes)
+			assert.Equal(t, cpum, cluster.Params.Replicaset.ComputeResources.CpuM)
+			//  assert.Equal(t, diskSize, cluster.Params.Replicaset.DiskSize)
```
Comment (BupycHuk) [inline tests/psmdbcluster/psmdb_cluster_test.go:91]: please pull changes from `pmm/2.13`

File: tests/xtradbcluster/xtra_db_cluster_test.go [hunk 3]
```diff
@@ -32,12 +32,36 @@ import (
 	"github.com/percona-platform/dbaas-controller/utils/app"
 )
 
+// func TestGetXtraDBClusterAPI(t *testing.T) {
```
Comment (BupycHuk) [inline tests/xtradbcluster/xtra_db_cluster_test.go:35]: let's uncomment or remove it

**Review by BupycHuk — CHANGES_REQUESTED** [review]
please pull changes from pmm/2.13 
move secret files from root directory into deploy directory and use them, 
update makefile to use yaml files from deploy directory

File: service/k8sclient/internal/pxc/pxc_types.go [hunk 4]
```diff
@@ -19,6 +19,7 @@
 
 import (
 	"github.com/percona-platform/dbaas-controller/service/k8sclient/internal/common"
+	corev1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply service/k8sclient/internal/pxc/pxc_types.go:22]: @percona-csalguero ^

**Review by BupycHuk — CHANGES_REQUESTED** [review]
please remove k8s dependencies

File: service/k8sclient/internal/pxc/pxc_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (percona-csalguero) [reply service/k8sclient/internal/pxc/pxc_types.go:22]: Should I create a new type?

File: deploy/delete_operators.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+PATH_TO_KUBECONFIG=${HOME}/.kube/config_eks
```
Comment (BupycHuk) [inline deploy/delete_operators.sh:3]: Let's move this logic into Makefile.
Also PATH_TO_KUBECONFIG should be configurable with default value.

File: deploy/install_eks_operators.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+PATH_TO_KUBECONFIG=${HOME}/.kube/config_eks
```
Comment (BupycHuk) [inline deploy/install_eks_operators.sh:3]: let's move this too to Makefile

File: service/cluster/psmdb_cluster.go [hunk 7]
```diff
@@ -183,6 +183,28 @@ func (s *PSMDBClusterService) RestartPSMDBCluster(ctx context.Context, req *cont
...
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer client.Cleanup() //nolint:errcheck
+
+	cluster, err := client.GetPSMDBCluster(ctx, req.Name)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	resp := &controllerv1beta1.GetPSMDBClusterResponse{
```
Comment (BupycHuk) [inline service/cluster/psmdb_cluster.go:199]: let's make our code more consistent and return all values here too.

File: service/k8sclient/internal/psmdb/psmdb_types.go [hunk 8]
```diff
@@ -168,9 +169,38 @@ type PodAffinity struct {
 	TopologyKey *string `json:"antiAffinityTopologyKey,omitempty"`
 }
 
+// ServiceType represents the expose type to open the cluster to the world.
+type ServiceType string
```
Comment (BupycHuk) [inline service/k8sclient/internal/psmdb/psmdb_types.go:173]: Let's extract it to common and not duplicate it in two places?

File: service/k8sclient/internal/pxc/pxc_types.go [hunk 9]
```diff
@@ -166,6 +189,7 @@ type PodSpec struct { //nolint:maligned
 	ServiceAccountName            string                          `json:"serviceAccountName,omitempty"`
 	ImagePullPolicy               common.PullPolicy               `json:"imagePullPolicy,omitempty"`
 	PodDisruptionBudget           *common.PodDisruptionBudgetSpec `json:"podDisruptionBudget,omitempty"`
+	ServiceType                   ServiceType
```
Comment (BupycHuk) [inline service/k8sclient/internal/pxc/pxc_types.go:192]: let's add `json` tag

File: service/k8sclient/k8client.go [hunk 10]
```diff
@@ -125,6 +125,7 @@ type XtraDBCluster struct {
 	Message  string
 	PXC      *PXC
 	ProxySQL *ProxySQL
+	Host     string
```
Comment (BupycHuk) [inline service/k8sclient/k8client.go:128]: let's remove it.

File: service/k8sclient/k8client.go [hunk 11]
```diff
@@ -313,6 +335,22 @@ func (c *K8Client) DeleteXtraDBCluster(ctx context.Context, name string) error {
...
+func (c *K8Client) GetXtraDBCluster(ctx context.Context, name string) (*XtraDBCredentials, error) {
+	var cluster pxc.PerconaXtraDBCluster
+
+	err := c.kubeCtl.Get(ctx, string(perconaXtraDBClusterKind), name, &cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get XtraDb cluster")
+	}
+
+	credentials := &XtraDBCredentials{
+		Host: cluster.Status.Host,
```
Comment (BupycHuk) [inline service/k8sclient/k8client.go:348]: let's return all values like we do for PSMDB.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Tests are failing

File: service/k8sclient/k8client.go (hunk 11 above)
Comment (percona-csalguero) [reply service/k8sclient/k8client.go:348]: There are no other credentials to return for the cluster.

Comment (BupycHuk) [reply service/k8sclient/k8client.go:403]: we can return static values from pmm-managed.

File: service/k8sclient/internal/psmdb/psmdb_types_test.go [hunk 12]
```diff
@@ -70,7 +70,8 @@ const expected = `
         "replsets": [
             {
                 "expose": {
-                    "enabled": false
+                    "enabled": false,
+					"exposeType": ""
```
Comment (BupycHuk) [inline service/k8sclient/internal/psmdb/psmdb_types_test.go:74]: let's use spaces instead of tabs.

File: tests/psmdbcluster/psmdb_cluster_test.go [hunk 13]
```diff
@@ -81,30 +86,43 @@ func TestPSMDBClusterAPI(t *testing.T) {
...
+			assert.Equal(t, memory, cluster.Params.Replicaset.ComputeResources.MemoryBytes)
+			assert.Equal(t, cpum, cluster.Params.Replicaset.ComputeResources.CpuM)
+			assert.Equal(t, diskSize, cluster.Params.Replicaset.DiskSize)
 			clusterFound = true
 		}
 	}
 	assert.True(t, clusterFound)
 
-	t.Log("Waiting for cluster to be ready")
+	t.Log("Wating for cluster to be ready")
```
Comment (BupycHuk) [inline tests/psmdbcluster/psmdb_cluster_test.go:98]:
```suggestion
	t.Log("Waiting for cluster to be ready")
```

File: tests/xtradbcluster/xtra_db_cluster_test.go [hunk 14]
```diff
@@ -62,17 +63,17 @@ func TestXtraDBClusterAPI(t *testing.T) {
 		},
 		Name: name,
 		Params: &controllerv1beta1.XtraDBClusterParams{
-			ClusterSize: 1,
```
Comment (BupycHuk) [inline tests/xtradbcluster/xtra_db_cluster_test.go:65]: is this changes required?

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Tests are failing because of
service/k8sclient/k8client_test.go:67:4: unknown field 'Expose' in struct literal of type XtraDBParams
service/k8sclient/k8client_test.go:127:4: unknown field 'Expose' in struct literal of type PSMDBParams

Please don't forget to run tests locally before passing to review.

File: tests/xtradbcluster/xtra_db_cluster_test.go (hunk 14 above)
Comment (percona-csalguero) [reply tests/xtradbcluster/xtra_db_cluster_test.go:65]: Set back to 2

File: Makefile [hunk 15]
```diff
@@ -15,6 +43,10 @@ PMM_RELEASE_VERSION ?=
 PMM_RELEASE_TIMESTAMP ?= $(shell date '+%s')
 PMM_RELEASE_FULLCOMMIT ?= $(shell git rev-parse HEAD)
 PMM_RELEASE_BRANCH ?= $(shell git describe --always --contains --all)
+PATH_TO_KUBECONFIG ?= ${HOME}/.kube/config_eks
```
Comment (askomorokhov) [inline Makefile:46]: I'd prefer to have as a default -  the default path to kubeconfig `${HOME}/.kube/config`

**Review by askomorokhov — APPROVED** [review]
Looks good to me with a minor comment.

File: service/k8sclient/kubectl/kubectl.go [hunk 16]
```diff
@@ -275,7 +275,7 @@ func run(ctx context.Context, kubectlCmd []string, args []string, stdin interfac
 		}
 	}
 
-	l.Debug(outBuf.String())
-	l.Debug(errBuf.String())
+	l.Infof(outBuf.String())
```
Comment (BupycHuk) [inline service/k8sclient/kubectl/kubectl.go:278]: let's revert it

File: tests/xtradbcluster/xtra_db_cluster_test.go (hunk 14 above)
Comment (BupycHuk) [reply tests/xtradbcluster/xtra_db_cluster_test.go:65]: I didn't get you, what do you mean?
what about CpuM?

Comment (percona-csalguero) [reply tests/xtradbcluster/xtra_db_cluster_test.go:65]: I've returned the clustersize to 1 and cpuM to 200 as the original test.

File: service/k8sclient/kubectl/kubectl.go [hunk 17]
```diff
@@ -275,7 +275,7 @@ func run(ctx context.Context, kubectlCmd []string, args []string, stdin interfac
 		}
 	}
 
-	l.Debug(outBuf.String())
-	l.Debug(errBuf.String())
+	l.Debugf(outBuf.String())
```
Comment (BupycHuk) [inline service/k8sclient/kubectl/kubectl.go:278]: please revert it.

**Review by BupycHuk — APPROVED** [review]

File: go.mod [hunk 18]
```diff
@@ -17,7 +17,7 @@ require (
 	github.com/google/uuid v1.1.2
 	github.com/grpc-ecosystem/go-grpc-middleware v1.2.2
 	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
-	github.com/percona-platform/dbaas-api v0.0.0-20201220195104-a35818be420f
+	github.com/percona-platform/dbaas-api v0.0.0-20201221131252-f97b6b79c864
```
Comment (AlekSi) [inline go.mod:20]: It was merged, please update

Comment (percona-csalguero) [reply go.mod]: Updated. All tests are passing.

**Review by AlekSi — APPROVED** [review]
